compiler/rustc_mir_dataflow/src/impls/live_borrows.rs

```diff
@@ -0,0 +1,642 @@
+        }
+
+        if num_succs == 0 {
+            // base node to keep alive
+            vec![src_node.data.get_local()]
+        } else {
+            if matches!(src_node.data, NodeKind::LocalOrLocalWithRefs(_)) {
+                // These are locals that we need to keep alive, but that also contain
+                // successors in the graph since they contain other references/pointers.
+                locals_for_node.push(current_local);
+            }
+
+            borrows_to_locals.insert(current_local, locals_for_node.clone());
+            locals_for_node
+        }
+    }
+}
+
+#[instrument(skip(tcx), level = "debug")]
+pub fn get_borrowed_locals_results<'mir, 'tcx>(
+    body: &'mir Body<'tcx>,
+    tcx: TyCtxt<'tcx>,
+) -> BorrowedLocalsResults<'mir, 'tcx> {
+    debug!("body: {:#?}", body);
+    let live_borrows = LiveBorrows::new(body, tcx);
+    let results =
+        live_borrows.into_engine(tcx, body).pass_name("borrowed_locals").iterate_to_fixpoint();
+
+    BorrowedLocalsResults::new(tcx, body, results)
+}
+
+pub struct BorrowedLocalsResultsCursor<'a, 'mir, 'tcx> {
+    body: &'mir Body<'tcx>,
+    borrows_analysis_cursor: ResultsRefCursor<'a, 'mir, 'tcx, LiveBorrows<'mir, 'tcx>>,
+    borrowed_local_to_locals_to_keep_alive: &'a FxHashMap<Local, Vec<Local>>,
+}
+
+impl<'a, 'mir, 'tcx> BorrowedLocalsResultsCursor<'a, 'mir, 'tcx> {
+    pub fn new(body: &'mir Body<'tcx>, results: &'a BorrowedLocalsResults<'mir, 'tcx>) -> Self {
+        let mut cursor = ResultsCursor::new(body, &results.borrows_analysis_results);
+
+        // We don't care about the order of the blocks, only about the result at a given location.
+        // This statement is necessary since we're performing a backward analysis in `LiveBorrows`,
+        // but want `Self::get` to be usable in forward analyses as well.
+        cursor.allow_unreachable();
+
+        Self {
+            body,
+            borrows_analysis_cursor: cursor,
+            borrowed_local_to_locals_to_keep_alive: &results.borrowed_local_to_locals_to_keep_alive,
+        }
+    }
+
+    /// Returns all `Local`s that need to be live at the given `Location` because of
+    /// outstanding references or raw pointers.
+    pub fn get(&mut self, loc: Location) -> BitSet<Local> {
+        self.borrows_analysis_cursor.seek_before_primary_effect(loc);
+        let live_borrows_at_loc = self.borrows_analysis_cursor.get();
+        debug!(?live_borrows_at_loc);
+
+        let mut borrowed_locals = BitSet::new_empty(live_borrows_at_loc.domain_size());
+        for borrowed_local in live_borrows_at_loc.iter() {
+            debug!(?borrowed_local);
+            if let Some(locals_to_keep_alive) =
+                self.borrowed_local_to_locals_to_keep_alive.get(&borrowed_local)
+            {
+                debug!(?locals_to_keep_alive);
+                for local in locals_to_keep_alive.iter() {
+                    borrowed_locals.insert(*local);
+                }
+            }
+        }
```

Comment:
It reaches a fixed point after one iteration. We perform the liveness analysis on the references and pointers and then use the dependency graph to infer for each borrowed local that is live at a given location which local we need to keep live at that location because of the outstanding borrow. We iterate the `LiveBorrows` analysis to a fixed point before initializing `BorrowedLocalsResults`.